src/distributed-transaction/percolator.md

```diff
@@ -0,0 +1,154 @@
+    * If there's already a lock or newer version than `start_ts`, the current transaction will be rolled back.
+
+And then, in the`Commit` phase:
+1. Get another timestamp, namely `commit_ts`.
+2. Remove the primary lock, and at the same time write a record to the `write` column with timestamp `commit_ts`, whose value records the transaction's `start_ts`. If the primary lock is missing, the commit fails.
+3. Repeat the process above for all secondary locks.
+
+Once step 2 (committing the primary) is done, the whole transaction is done. It doesn't matter if the process of committing the secondary locks failed.
+
+Let's see the example from the paper of Percolator. Assume we are writing two rows in a single transaction. At first, the data looks like this:
+
+| key | bal:data     | bal:lock  | bal:write       |
+|-----|--------------|-----------|-----------------|
+| Bob | 6:<br/>5:$10 | 6:<br/>5: | 6:data@5<br/>5: |
+| Joe | 6:<br/>5:$2  | 6:<br/>5: | 6:data@5<br/>5: |
+
+This table shows Bob and Joe's balance. Now Bob wants to transfer his $7 to Joe's account. The first step is `Prewrite`:
+
+1. Get the `start_ts` of the transaction. In our example, it's `7`.
+2. For each row involved in this transaction, put a lock in the `lock` column, and write the data to the `data` column. One of the locks will be chosen as the primary lock.
+
+After `Prewrite`, our data looks like this:
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 7:$3<br/>6:<br>5:$10 | 7:primary<br/>6:<br/>5:         | 7:<br/>6:data@5<br/>5: |
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+Then `Commit`:
+
+1. Get the `commit_ts`, in our case, `8`.
+2. Commit the primary: Remove the primary lock and write the commit record to the `write` column.
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 7:$9<br/>6:<br/>5:$2 | 7:primary@Bob.bal<br/>6:<br/>5: | 7:<br/>6:data@5<br/>5: |
+
+3. Commit all secondary locks to complete the writing process.
+
+| key | bal:data | bal:lock | bal:write |
+|-----|----------|----------|-----------|
+| Bob | 8:<br/>7:$3<br/>6:<br>5:$10 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+| Joe | 8:<br/>7:$9<br/>6:<br/>5:$2 | 8:<br/>7:<br/>6:<br/>5: | 8:data@7<br/>7:<br/>6:data@5<br/>5: |
+
+### Reading
+
+Reading from Percolator also requires a timestamp. The procedure to perform a read operation is as follows:
+
+1. Get a timestamp, `ts`.
+2. Check if the row we are going to read is locked with a timestamp in the range `[0, ts]`;
+    * If there is a timestamp in range `[0, ts]`, it means the row is locked by an earlier-started transaction. Then we are not sure whether that transaction will be committed before or after `ts`.
```

Comment:
Do we need to mention how to deal with this situation? e.g. back off or clean up the lock?